src/mmfatl.c

```diff
@@ -0,0 +1,1591 @@
+ * sequence may not be transferred completely.  Trace back to the last
+ * character of a completely transferred sequence.  ASCII encoding is a single byte
+ * subset encoding of UTF-8 that always yields the last transferred byte.
+ * \param begin [not null] a pointer to the buffer start.
+ * \param last [not null] a pointer to the last transferred byte in the buffer.
+ * \returns a pointer to the last valid byte, or NULL if it does not exist.
+ *   The latter can happen only if either your buffer is less than 4 bytes
+ *   long, or if it contains invalid UTF-8 encoding.
+ */
+
+static char* findLastUTF8Byte(char const* begin, char* end)
+{
+    if (*end >= 0) // ASCII character
+        return end;
+    char* pt = end + 1;
+    // skip backwards all UTF-8 continuation bytes
+    while ((*--pt & 0xC0) == 0x80 && pt != begin);
+    // should be a leading byte here, otherwise invalid UTF-8
+    unsigned char leadByte = *pt;
+    if ((leadByte & 0xC0) <= 0x80)
+        return NULL;
+    // number of stored continuaton bytes
+    int buffered = end - pt;
+    if (buffered > 3)
+        return NULL;
+    // up to 3 continuation bytes possible
+    int expected = leadByte >= 0xF0? 3 : (leadByte >= 0xE0? 2 : 1);
+    return buffered == expected? end : (pt == begin? NULL : pt - 1);
+}
+
+/*=============   Allocation of Memory for Fatal Error Messages   ===========*/
+
+/*!
+ * a \ref FatalErrorBufferDescriptor describing a pre-allocated
+ * buffer for fatal error messages.  The initial settings indicate the absence
+ * of such a buffer, which is allowed during program startup only.  Should be
+ * modified by \ref allocFatalErrorBuffer
+ */
+static struct FatalErrorBufferDescriptor descriptor = {
+    0,  /* size */
+    0,  /* dmz */
+    NULL, /* no ellipsis.  Null only accepted during startup */
+};
+
+/*!
+ * pre-allocated memory block used for fatal error messages.  The size of the
+ * block is derived from a \ref FatalErrorBufferDescriptor.  Only a single
+ * global instance is maintained.  All values assigned to \p memBlock must be a
+ * result of a \p malloc call.
+ * \invariant \ref descriptor describes the memory layout correctly if this
+ *   pointer is different from NULL.
+ */
+static void* memBlock = NULL;
+
+struct FatalErrorBufferDescriptor const* getFatalErrorBufferDescriptor()
+{
+    return &descriptor;
+}
+
+/*!
+ * does the submitted \ref FatalErrorBufferDescriptor fulfil minimum
+ * requirements for a valid pre-allocation?
+ * \attention an affirmative result is not the final answer.  Allocation
+ *   requires also information about available memory.
+ * \param aDescriptor [null] a pointer to a \ref FatalErrorBufferDescriptor
+ *   under investigation.
+ * \returns 1 if acceptable, 0 else.
+ */
+static int isValidFatalErrorBufferDescriptor(
+    struct FatalErrorBufferDescriptor const* aDescriptor)
+{
+    return aDescriptor != NULL
+        && aDescriptor->size > 0
+        && aDescriptor->ellipsis != NULL? TRUE : FALSE;
```

Comment:
I can't tell from this indentation whether you want everything in the conditional or whether the last conjunct is supposed to be `aDescriptor->ellipsis != NULL? TRUE : FALSE`. Also `x ? TRUE : FALSE` is bonkers, that's just `x`. (Everything here is a proper boolean value so it can't be that you want to convert an int to a bool.)

Reply:
Not exactly.  C discriminates between 0 and nonzero, but not between 0 and 1.

Reply:
What does that mean? The relational operators [are defined to always return 0 or 1](https://stackoverflow.com/a/37292153/890016).

Reply:
I did not know that.